beginner_source/transformer_tutorial.py

```diff
@@ -0,0 +1,326 @@
+"""
+nn.Transformer Tutorial
+============================
+
+This is a tutorial to show how to implement `nn.Transformer <https://pytorch.org/docs/master/nn.html?highlight=nn%20transformer#torch.nn.Transformer>`__ module.
+
+PyTorch 1.2 release includes a standard transformer module based on the
+paper `Attention is All You
+Need <https://arxiv.org/pdf/1706.03762.pdf>`__. The transformer model
+has been proved to be superior in quality for many sequence-to-sequence
+problems while being more parallelizable. The ``nn.Transformer`` module
+relies entirely on an attention mechanism (another module recently
+implemented as `nn.MultiheadAttention <https://pytorch.org/docs/master/nn.html?highlight=multiheadattention#torch.nn.MultiheadAttention>`__) to draw global dependencies
+between input and output. The ``nn.Transformer`` module is now highly
+modularized such that a single component (like `nn.TransformerEncoder <https://pytorch.org/docs/master/nn.html?highlight=nn%20transformerencoder#torch.nn.TransformerEncoder>`__
+in this tutorial) can be easily adapted/composed.
+
+.. image:: ../_static/img/transformer_architecture.jpg
+
+"""
+
+######################################################################
+# Define the model
+# ----------------
+# 
+
+
+######################################################################
+# In this tutorial, we train ``nn.TransformerEncoder`` model on a
+# language modeling task. The language modeling task is to assign a
+# probability for the likelihood of a given word (or a sequence of words)
+# to follow a sequence of words. A sequence of tokens are passed to the embedding
+# layer first, followed by a positional encoding layer to account for the order
+# of the word (see the next paragraph for more details). The
+# ``nn.TransformerEncoder`` consists of multiple layers of
+# `nn.TransformerEncoderLayer <https://pytorch.org/docs/master/nn.html?highlight=transformerencoderlayer#torch.nn.TransformerEncoderLayer>`__. Along with the input sequence, a square
+# attention mask is required because the self-attention layers in
+# ``nn.TransformerEncoder`` are only allowed to attend the earlier positions in
+# the sequence. For the language modeling task, any tokens on the future
+# positions should be masked. To have the actual words, the output
+# of ``nn.TransformerEncoder`` model is sent to the final Linear
+# layer, which is followed by a log-Softmax function.
+# 
+
+import math
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+class TransformerModel(nn.Module):
+
+    def __init__(self, ntoken, ninp, nhead, nhid, nlayers, dropout=0.5):
+        super(TransformerModel, self).__init__()
+        from torch.nn import TransformerEncoder, TransformerEncoderLayer
```

Comment:
Is there a reason this import is inside the `Module`?

Reply:
not really. I could move them out of the class.